public/app/features/alerting/unified/rule-list/GroupedView.tsx

```diff
@@ -56,223 +47,52 @@ export function DataSourceLoader({ uid, name }: DataSourceLoaderProps) {
-            title={
-              <Stack direction="row" gap={1} alignItems="center">
-                <Icon name="folder" />{' '}
-                <Text variant="body" element="h3">
-                  {namespace.name}
-                </Text>
-              </Stack>
-            }
-          >
-            {namespace.groups.map((group) => (
-              <RuleGroupListItem
-                key={`${ruleSourceName}-${namespace.name}-${group.name}`}
-                group={group}
-                ruleSourceName={ruleSourceName}
-                namespaceId={namespace}
-              />
-            ))}
-          </ListSection>
-        ))}
-        <LazyPagination
-          nextPage={nextPage}
-          previousPage={previousPage}
-          canMoveForward={canMoveForward}
-          canMoveBackward={canMoveBackward}
-        />
-      </Stack>
-    </DataSourceSection>
-  );
-}
-
-interface RuleGroupListItemProps {
-  group: RuleGroup;
-  ruleSourceName: string;
-  namespaceId: DataSourceNamespaceIdentifier;
-}
-
-function RuleGroupListItem({ group, ruleSourceName, namespaceId }: RuleGroupListItemProps) {
-  const rulesWithGroupId = useMemo(
-    () =>
-      group.rules.map((rule) => {
-        const groupIdentifier: DataSourceRuleGroupIdentifier = {
-          rulesSource: { uid: getDatasourceAPIUid(ruleSourceName), name: ruleSourceName },
-          namespace: namespaceId,
-          groupName: group.name,
-          groupOrigin: 'datasource',
-        };
-        return { rule, groupIdentifier };
-      }),
-    [group, namespaceId, ruleSourceName]
-  );
-
-  return (
-    <ListGroup
-      key={group.name}
-      name={group.name}
-      isOpen={false}
-      actions={
-        <>
-          <Dropdown
-            overlay={
-              <Menu>
-                <Menu.Item label="Edit" icon="pen" data-testid="edit-group-action" />
-                <Menu.Item label="Re-order rules" icon="flip" />
-                <Menu.Divider />
-                <Menu.Item label="Export" icon="download-alt" />
-                <Menu.Item label="Delete" icon="trash-alt" destructive />
-              </Menu>
-            }
-          >
-            <IconButton name="ellipsis-h" aria-label="rule group actions" />
-          </Dropdown>
-        </>
-      }
-    >
-      {rulesWithGroupId.map(({ rule, groupIdentifier }) => (
-        <AlertRuleLoader key={hashRule(rule)} rule={rule} groupIdentifier={groupIdentifier} />
-      ))}
-    </ListGroup>
-  );
-}
-
-interface DataSourceSectionProps extends PropsWithChildren {
-  uid: string;
-  name: string;
-  loader?: ReactNode;
-  application?: RulesSourceApplication;
-  isLoading?: boolean;
-  description?: ReactNode;
-}
-
-const DataSourceSection = ({
-  uid,
-  name,
-  application,
+/**
+ * Some more exotic Prometheus data sources might not be 100% compatible with Prometheus API
+ * We don't want them to break the whole page, so we wrap them in an error boundary
+ */
+function DataSourceErrorBoundary({
```

Comment:
I like this a lot, nice one :+1: What's the easiest way to test this scenario out locally?
Could we move this to a separate file?

Reply:
To test I was overriding the server response in the browser. This is mainly intended for situations where the API is not 100% compatible with Prometheus and the server-provided object might not be accurate with our types